### PR TITLE
Add numeric conversions

### PR DESCRIPTION
**Pull request by stephentetley**
Flix is currently missing conversion functions between its number types, this pull request adds them.

The numeric conversions in this PR are "basic" but possibly not "simple" - number conversion turns out to be a bit tricky.

There are conversions for all seven of Flix's number types.

The overriding principle in the implementation / API was to avoid overflow - if any numeric values allowed in the source type cannot be represented in the target type we wrap the target type with Option and unrepresentable values become `None`. This means the API is not "symmetric" but it does mean it can avoid overflow and/or exceptions.

The implementation makes no attempt to address loss of precision when changing to values that are representable in the target type. Loss of precision is an intrinsic problem when using "machine level" numeric types.

As well as the number conversion functions I've added `toDecimalString` to Float32 and Float64 so I could get an idea of how much loss of precision was going on (I found this difficult with `toString` as it uses scientific notation for all but the simplest values). I like having this function, but maybe the current output format is not the nicest to have as a default.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Thanks Stephen!

This is some serious work. 

I think the design looks very reasonable (and it does not prevent us from complementing/extending it later). I want to read up a bit on machine integers since this is a topic I really know almost nothing about.. so my review might be a bit later than usual. Plausibly there will be more comments in #1059.

**Comment by stephentetley** [comment]
Thanks Magnus. There's no particular rush, I've still got plenty to work on with the string-additions.

File: main/src/library/BigInt.flix [hunk 1]
```diff
@@ -97,6 +97,116 @@ namespace BigInt {
...
+        let d = xd.floatValue() as & Pure;
+        if (Float32.isInfinite(d) || Float32.isNan(d))
+            None
+        else
+            let* f32min = Float32.toBigInt(Float32.minValue());
+            let* f32max = Float32.toBigInt(Float32.maxValue());
+            if (x < f32min || x > f32max)
+                None
+            else
+                Some(d)
```
Comment (magnus-madsen) [inline main/src/library/BigInt.flix:183]: I think I understand the last part. You ensure that the number is not too big or small to fit within a float. What does the first part do? How can the number become infinite or NaN?  Looking at the JavaDoc for floatValue I see that it can return infinity or negative infinity. I guess that `isInfinite` checks for both. But what about NaN? 

But wait, if floatValue() can return infinity, why do we need the latter checks? Is something more subtle going on?

Comment (magnus-madsen) [reply main/src/library/BigInt.flix:183]: Side note: I guess we should add support for BigDecimal at some point.

File: main/src/library/Float32.flix [hunk 2]
```diff
@@ -104,4 +110,91 @@ namespace Float32 {
...
+    /// Convert `x` to an `Option[Int8]`.
+    ///
+    /// If the numeric value of `x` is within the range of Int8 return `Some(x as Int8)`, rounding `x` towards 0`.
+    ///
+    /// If the numeric value of `x` is outside the range of Int8, or it is NaN or infinity, return `None`.
+    ///
+    pub def toInt8(x: Float32): Option[Int8] =
+        import java.lang.Float:valueOf(Float32);
+        import java.lang.Float.byteValue();
+        if (x < Int8.toFloat32(Int8.minValue()) || x > Int8.toFloat32(Int8.maxValue()) || isNan(x) || isInfinite(x))
```
Comment (magnus-madsen) [inline main/src/library/Float32.flix:124]: OK; I think I understand this.

I was wondering if we should have some operation where you can supply a value for the "negative infinity" and "positive infinity" cases and/or where they are automatically "clamped" to Int8.minValue() and Int8.maxValue(). Kind of like how we have `Option.getWithDefault`.

Comment (magnus-madsen) [reply main/src/library/Float32.flix:124]: Maybe like `toInt8WithMinMax` and/or `toInt8With(min = ..., max = ...)`

File: main/src/library/BigInt.flix (hunk 1 above)
Comment (stephentetley) [reply main/src/library/BigInt.flix]: The inifinity checks are redundant - because d was originally a BigInt it would not get converted to either -Infinity or +Infinity. I've pushed a change to fix this.

BigDecimal support would be good.

File: main/src/library/BigInt.flix [hunk 3]
```diff
@@ -97,6 +97,116 @@ namespace BigInt {
         case _: ##java.lang.NumberFormatException => Err("BigInt.fromString")
     }
 
+    ///
+    /// Convert `x` to an `Option[Int8]`.
```
[the remark this answers is not in the document]
Comment (stephentetley) [reply main/src/library/BigInt.flix:101]: F# favours using `try` as prefix for things it can't do with a total function and has to return an Option (e.g. tryHead). 

I'd be quite happy with this convention in "moderation" e.g. `head` and `last` in Flix's List module are safe versions that return `Option[a]` . I don't think they should be renamed to `tryHead` and `tryLast` but for these number conversions `tryToInt8` would be fine.

File: main/src/library/BigInt.flix [hunk 4]
```diff
@@ -97,6 +97,116 @@ namespace BigInt {
...
+    ///
+    /// If the numeric value of `x` can be represented exactly return `Some(x as Int8)`.
+    ///
+    /// If the numeric value of `x` is outside the range of Int8 return `None`.
```
[the remark this answers is not in the document]
Comment (stephentetley) [reply main/src/library/BigInt.flix]: Ranges now added in comments.

File: main/src/library/BigInt.flix [hunk 5]
```diff
@@ -97,6 +97,116 @@ namespace BigInt {
         case _: ##java.lang.NumberFormatException => Err("BigInt.fromString")
     }
 
+    ///
+    /// Convert `x` to an `Option[Int8]`.
+    ///
+    /// If the numeric value of `x` can be represented exactly return `Some(x as Int8)`.
```
[the remark this answers is not in the document]
Comment (stephentetley) [reply main/src/library/BigInt.flix]: I like that style too, so I've changed the comments.

File: main/src/library/Float32.flix (hunk 2 above)
Comment (stephentetley) [reply main/src/library/Float32.flix:126]: I like the idea of supplying a min and max value for clamping, because the min and max values will be in the target type you won't be able to mix up the arg order if you you thinking of UFCS or "not UFCS". 

I think I'd prefer to have "clamp" in the name though, rather than say `toInt8WithMinMax`,  this would stop flag to a new user that the clamp version shouldn't be seen as the default choice for a conversion function.

Because we have `Option.getWithDefault` in UFCS order I'll add functions like this:

~~~

pub def clampToInt8(x: Int64, min: Int8, max: Int8): Int8 = 

~~~

File: main/src/library/BigInt.flix (hunk 3 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/BigInt.flix:101]: I think this sounds like a good compromise.

File: main/src/library/Float32.flix (hunk 2 above)
Comment (magnus-madsen) [reply main/src/library/Float32.flix:126]: Great!

File: main/src/library/BigInt.flix (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephentetley) [reply main/src/library/BigInt.flix:101]: Good. I'll change the "Option" ones to `try_` when I add the `clamp_` functions.

[the remark this answers is not in the document]
Comment (stephentetley) [reply main/src/library/BigInt.flix:101]: Name changes to use `try` prefix have been pushed.

File: main/src/library/Float32.flix (hunk 2 above)
Comment (stephentetley) [reply main/src/library/Float32.flix:126]: I've push the clamp functions but the tests are either incomplete or missing.

**Comment by stephentetley** [comment]
I'm still working on the clamp functions - all are in place but the tests need writing (where I've written some of the tests, I think I still need to cover some boundary cases).

I'm back at work tomorrow - I've been on holiday for a fortnight which is how I've been so productive - though I plan to work on the tests in the evening.

**Comment by stephentetley** [comment]
Hi Magnus, I've added the clamp functions, they are either the nice 3 argument form or a less nice 4 argument version if they have to deal with NaN:

~~~

pub def clampToFloat32(x: BigInt, minimum: Float32, maximum: Float32): Float32 =

pub def clampToInt64(x: Float64, minimum: Int64, maximum: Int64, nanValue: Int64): Int64 =

~~~

Given the constraint that we want these functions to return a number and not wrap it in an Option, what should we do about bad ranges? i.e when minimum as greater than maximum?

I'm inclined to always return 0 from the clamping function in this case (and put in warning in the comment that this is the behavior). I suspect this will be easier to find through debugging rather than returning either the minimum or the maximum.


What do you think? Thanks

**Comment by magnus-madsen** [comment]
> I'm inclined to always return 0 from the clamping function in this case (and put in warning in the comment that this is the behavior). I suspect this will be easier to find through debugging rather than returning either the minimum or the maximum.

Hmm. This is a hard question. Is there any reason not be pedantic and simply to what you tell us? I.e. if you give us any minimum value then any value too small gets clamped to that. No matter what maximum is.

File: main/src/library/BigInt.flix [hunk 6]
```diff
@@ -97,6 +97,234 @@ namespace BigInt {
...
+    ///
+    pub def tryToFloat32(x: BigInt): Option[Float32] =
+        use Option.flatMap;
+        import new java.math.BigDecimal(BigInt) as newBigDecimal;
+        import java.math.BigDecimal.floatValue();
+        let xd = newBigDecimal(x) as & Pure;
+        let d = xd.floatValue() as & Pure;
+        let* f32min = Float32.tryToBigInt(Float32.minValue());
+        let* f32max = Float32.tryToBigInt(Float32.maxValue());
+        if (x < f32min || x > f32max)
```
Comment (magnus-madsen) [inline main/src/library/BigInt.flix:182]: I was tought to write such limits as:

```
if (f32min <= x && x <= f32max) 
  Some(d)
```
because it closely mirrors mathematical syntax of `x < y < z`.

**Comment by magnus-madsen** [comment]
I went ahead and merged this. I think we can deal with the remaining issues in separate PRs.
